rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/FileMetadata.java

```diff
@@ -41,30 +48,68 @@
   private long size;
   private long lastModified;
   private long currentFileSize;
-
   private boolean needsRefresh = true;
+  private int generation;
 
   int bitmapFileSizeBytes;
   ByteBufferBitmap blockBitmap;
 
   static Striped<Lock> stripes = Striped.lock(20000);
+  //  Maintains generation number for remote file
+  //  If multiple threads call get of guava cache , then all are blocked and computation is done for one
+  //  and values is returned to other blocked threads.
+  static Cache<String, Integer> generationNumber = CacheBuilder.newBuilder()
+          .expireAfterWrite(2, TimeUnit.HOURS)
+          .build();
+  static BloomFilter filter = BloomFilter.create(Funnels.stringFunnel(Charset.defaultCharset()), 1000000, 0.01);;
 
   private static Log log = LogFactory.getLog(FileMetadata.class.getName());
 
   public FileMetadata()
   {
   }
 
-  public FileMetadata(String remotePath, long fileLength, long lastModified, long currentFileSize, Configuration conf)
-      throws IOException
+  public FileMetadata(String remotePath, long fileLength, long lastModified, long currentFileSize, Configuration conf, boolean newFileEntry)
+          throws IOException, ExecutionException
   {
     this.remotePath = remotePath;
     this.size = fileLength;
     this.lastModified = lastModified;
     this.currentFileSize = currentFileSize;
-    localPath = CacheUtil.getLocalPath(remotePath, conf);
-    mdFilePath = CacheUtil.getMetadataFilePath(remotePath, conf);
-
+    Lock lock = stripes.get(remotePath);
+    try {
+      lock.lock();
+      if (newFileEntry) {
+        int genNumber = 0;
+        if (!filter.mightContain(remotePath)) {
+          // BKS restarted
+          while (new File(CacheUtil.getLocalPath(remotePath, conf, genNumber)).exists()) {
+            genNumber++;
+          }
+          if (genNumber != 0) genNumber--;
+        }
+        else {
+            genNumber = generationNumber.get(remotePath, new Callable<Integer>() {
+              @Override
+              public Integer call() throws Exception {
+                return -1;
+              }
+            });
+            genNumber++;
+            while (new File(CacheUtil.getLocalPath(remotePath, conf, genNumber)).exists()) {
```

Comment:
why is this logic needed if we are reading from cache? Also, cache may return '-1' if it doesn't have the path, we should add a check.

Reply:
This is just a failsafe, in case there are files existing already. This ideally should never happen